File: src/codes/duplicate-logic/ts/rfc_prototype.ts
interface ExporterPrototype {
  clone(): ExporterPrototype;
  export(data: string): void;
}

class PdfExporter implements ExporterPrototype {
  constructor(private withLog: boolean = false) {}

  clone(): ExporterPrototype {
    return new PdfExporter(this.withLog);
  }

  export(data: string): void {
    if (this.withLog) console.log("開始ログ");
    console.log(`PDF出力: ${data}`);
    if (this.withLog) console.log("完了ログ");
  }

  enableLogging(): void {
    this.withLog = true;
  }
}

// 利用例
const baseExporter = new PdfExporter();
baseExporter.enableLogging();

const cloned = baseExporter.clone();
cloned.export("請求書");

File: src/codes/duplicate-logic/py/rfc_prototype.py
import copy
from abc import ABC, abstractmethod

class ExporterPrototype(ABC):
    @abstractmethod
    def clone(self): pass

    @abstractmethod
    def export(self, data: str): pass

class PdfExporter(ExporterPrototype):
    def __init__(self, with_log: bool = False):
        self.with_log = with_log

    def clone(self) -> "PdfExporter":
        return copy.deepcopy(self)

    def enable_logging(self):
        self.with_log = True

    def export(self, data: str):
        if self.with_log:
            print("開始ログ")
        print(f"PDF出力: {data}")
        if self.with_log:
            print("完了ログ")

# 利用例
base_exporter = PdfExporter()
base_exporter.enable_logging()

cloned = base_exporter.clone()
cloned.export("請求書")
